### DISCUSSION TITLE
Looking for a way to detect stalled boot

### DISCUSSION THREAD

**Discussion by maburlik** [cloud-hypervisor/cloud-hypervisor#6278] [General]
I'm running in an environment where I've experienced occasional stalls/hangs during successive cloud-hypervisor VM create & boots and am trying to come up with automation on the host that would detect when the VM boot abruptly halts. In particular, I've seen that the VM early syslogs are emitted via stdout, then abruptly terminate within the first few seconds.

It would be nice to know what in particular has stalled, and whether the kernel is still operating. Has the kernel panicked, is systemd stuck at a dependency step?

Currently I have some limited monitoring capabilities on the host to inspect cloud-hypervisor state via the REST api, as well as an agent in the guest that sends messages via vsock. It is the intermediate phase between start of boot and first message via the agent where I am missing information.

I've just found out about the event-monitor capabilities and still need to run some experiments to find what data I can get, and whether this can be leveraged by automation to help determine a VM stall. What I'm looking to get is effectively a liveness monitoring capability that would indicate whether the VM has gone through recent compute cycles in the guest.

I've considered adding inspection hooks on cloud-hypervisor stdout but given that I want to respond to state changes via automation, I'd prefer to take a dependency on well-formed data such as a vm.info state change, trigger, or event stream.

Does cloud-hypervisor have any liveness monitoring capabilities of the guest? Is there any data to which it has access, that could be used to infer liveness?

**Comment by rbradford** [comment]
Would our virtio-watchdog implementation work for you? If the watchdog does not receive a notification from userspace (via systemd) it will reboot the VM.

**Reply by liuw** [reply]
virtio-watchdog is probably the best bet.

**Reply by maburlik** [reply]
I'll look into it. Thanks for the suggestion!
